### ISSUE TITLE
`aten::_trilinear` returns zeros or zero-sized dimensions when all inputs expand the same dimension

### ISSUE THREAD

**Issue by Tongkaio** [pytorch/pytorch#196424]
### 🐛 Describe the bug

`aten::_trilinear` produces unexpected results when `i1`, `i2`, and `i3` all insert a singleton dimension at the same logical position.

The implementation comment describes the operation as equivalent to:

```python
(i1.unsqueeze(expand1) * i2.unsqueeze(expand2) * i3.unsqueeze(expand3)).sum(sumdim)
```

A dimension inserted by all three inputs has size 1 according to this formula. However, the native implementation treats its size as 0. This causes two observable problems:

1. If the shared expanded dimension is `unroll_dim`, the unroll loop executes zero times and returns an all-zero tensor.
2. If the shared expanded dimension is not in `sumdim`, the output contains a zero-sized dimension instead of a singleton dimension.

This is related to #138353, which reports the second problem. This report also covers the all-zero result and explains the common root cause.

### Minimal reproduction

```python
import torch


def formula(i1, i2, i3, expand1, expand2, expand3, sumdim):
    for dim in sorted(expand1):
        i1 = i1.unsqueeze(dim)
    for dim in sorted(expand2):
        i2 = i2.unsqueeze(dim)
    for dim in sorted(expand3):
        i3 = i3.unsqueeze(dim)

    result = i1 * i2 * i3
    for dim in sorted(sumdim, reverse=True):
        result = result.sum(dim)
    return result


i1 = torch.ones(2, 3)
i2 = torch.ones(2, 3)
i3 = torch.ones(2, 3)

# Problem 1: the dimension expanded by all inputs is also unroll_dim.
actual = torch.ops.aten._trilinear.default(
    i1, i2, i3,
    [0], [0], [0],
    [0, 2],
    0,
)
expected = formula(i1, i2, i3, [0], [0], [0], [0, 2])

print("Problem 1")
print("actual:", actual.shape, actual)
print("expected:", expected.shape, expected)

# Problem 2: the dimension expanded by all inputs is not reduced.
actual = torch.ops.aten._trilinear.default(
    i1, i2, i3,
    [0], [0], [0],
    [2],
    1,
)
expected = formula(i1, i2, i3, [0], [0], [0], [2])

print("Problem 2")
print("actual:", actual.shape, actual)
print("expected:", expected.shape, expected)

# Control: a shared expanded dimension that is reduced but is not unroll_dim.
actual = torch.ops.aten._trilinear.default(
    i1, i2, i3,
    [0], [0], [0],
    [0, 2],
    1,
)
expected = formula(i1, i2, i3, [0], [0], [0], [0, 2])

print("Control")
print("actual:", actual.shape, actual)
print("expected:", expected.shape, expected)
```

### Actual output

```text
Problem 1
actual: torch.Size([2]) tensor([0., 0.])
expected: torch.Size([2]) tensor([3., 3.])

Problem 2
actual: torch.Size([0, 2]) tensor([], size=(0, 2))
expected: torch.Size([1, 2]) tensor([[3., 3.]])

Control
actual: torch.Size([2]) tensor([3., 3.])
expected: torch.Size([2]) tensor([3., 3.])
```

### Expected behavior

For a dimension inserted by all three inputs, its logical size should be 1.

- Problem 1 should execute one unroll iteration and return `tensor([3., 3.])`.
- Problem 2 should return a tensor with shape `[1, 2]`, not `[0, 2]`.

This is not a numerical-accuracy or associativity difference: the first result is exactly zero instead of three, and the second result has a different shape.

### Root cause

The issue appears to originate in `aten/src/ATen/native/Linear.cpp` inside `_trilinear`：

https://github.com/pytorch/pytorch/blob/de1b9032f48b5e91b14b39fc07c29456ce1da6d4/aten/src/ATen/native/Linear.cpp?spm=21540d8c.363b7dbd.0.0.6212c9cassg3n6#L678-L752

When all three inputs expand dimension `i`, none of the three `else` branches assigns a tensor size to `s`, so it remains 0.

The zero is then used in two ways:

- `unroll_size = 0` causes the accumulation loop to execute zero times, leaving the preallocated output filled with zeros.
- `output_size.push_back(s)` creates a zero-sized output dimension when that dimension is not reduced.

Initializing `s` to 1 may match the singleton dimension introduced by `unsqueeze`, while still allowing a non-expanded input to overwrite it with its actual size. Additional validation may be appropriate if this parameter combination is considered unsupported.

### Versions

PyTorch version: 2.9.0

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano

**Comment by Mramaran** [comment]
I ran the repro against current main and can confirm both problems.

The cause is in `_trilinear` in `aten/src/ATen/native/Linear.cpp`. The loop over dimensions tracks the size of dimension `i` with `c10::SymInt s = 0;`, and each of the three inputs only assigns `s` in its non-expand branch (`s = i1.sym_size(i)` etc.). When `i1`, `i2` and `i3` all expand the same dimension, none of those branches run, so `s` keeps its initial value of 0. That zero then reaches two places:

1. `output_size.push_back(sumdim[i] ? 1 : s)` records a 0-sized output dimension when the shared dimension is not reduced. That is problem 2, and the same thing #138353 reports.
2. `unroll_size = s.guard_int(...)` becomes 0 when the shared dimension is `unroll_dim`. The unroll loop never runs and the result is left at its `at::zeros` initial value. That is problem 1.

The control case in the report works because the shared dimension is both reduced and not unrolled: the `sumdim[i] ? 1 : s` branch hides the zero, and `sumproduct_pair` sums over a size-1 dimension as expected.

The comment above the function describes the op as `(i1.unsqueeze(...) * i2.unsqueeze(...) * i3.unsqueeze(...)).sum(sumdim)`, and an unsqueezed dimension has size 1, so the initial value should be `c10::SymInt s = 1;`. Whenever any input does not expand dimension `i`, `s` is overwritten with that input's real size, so the change only affects the all-expand case. Tracing problem 1 with `s = 1`: `unroll_size` is 1, the loop runs once, and the output is a `(2,)` tensor of 3s. Problem 2 gives shape `(1, 2)`.

I have not built the change yet, only traced it by hand. I can send a PR with the one-line fix and a regression test covering both cases if a maintainer marks this actionable.
